File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
from collections import deque

def solution(numbers, target):
    answer = 0
    q = deque([(0,0)])
    
    while q:
        idx, curr_sum = q.popleft()
        
        # 종료 조건
        if idx == len(numbers):
            if curr_sum == target:
                answer += 1
        else:
            # 다음 상태 전이
            q.append((idx+1, curr_sum + numbers[idx]))
            q.append((idx+1, curr_sum - numbers[idx]))
                
    return answer

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    var answer = 0
    
    fun dfs(idx: Int, sum: Int, numbers: IntArray, target: Int){
        // 종료 조건
        if(idx >= numbers.size -1){
            if(sum == target){
                answer++
            }
            return
        }
        
        
        dfs(idx+1, sum + numbers[idx+1], numbers, target)
        dfs(idx+1, sum - numbers[idx+1], numbers, target)
        
        
    }
    
    fun solution(numbers: IntArray, target: Int): Int {
        dfs(-1, 0, numbers, target)
        
        return answer
    }
}